File: server/src/main/java/com/example/backend/modules/room/services/RoomService.java
package com.example.backend.modules.room.services;

import com.example.backend.commons.ResponsePaging;
import com.example.backend.commons.ResponseSuccess;
import com.example.backend.modules.room.dtos.CreateRoomDTO;
import com.example.backend.modules.room.dtos.EditRoomDTO;
import com.example.backend.modules.room.models.Room;
import com.example.backend.modules.room.viewmodels.RoomVm;
import com.example.backend.modules.user.models.User;
import com.example.backend.modules.user.viewmodels.UserVm;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public interface RoomService {

    Optional<Room> findById(int id);

    boolean isRoomOwner(User user, int roomId);

    long countUserInRoom(Room room);

    ResponseSuccess<Integer> joinRoom(User user,String roomPin);

    ResponseSuccess<RoomVm> getRoomForParticipants(User user,String roomPin);

    ResponseSuccess<RoomVm> createRoom(User user, CreateRoomDTO dto);
    ResponseSuccess<Boolean> endRoom(User user,int roomId);

    ResponseSuccess<Boolean> editRoom(User user, EditRoomDTO dto);
    ResponseSuccess<Boolean> deleteRoom(User user,int roomId);

    ResponseSuccess<ResponsePaging<List<RoomVm>>> getMyListRooms(String keyword,String sortBy,int pageIndex,User user); // phân trang

    ResponseSuccess<RoomVm> getRoomDetail(int roomId);

    ResponseSuccess<ResponsePaging<List<RoomVm>>> getJoinedRoom(User user,int pageIndex);
    ResponseSuccess<Boolean> kickUser(User user, int roomId, int userId);

    ResponseSuccess<ResponsePaging<List<UserVm>>> getUsersInRoom(int roomId,int pageIndex);
}

File: client/app/src/main/java/com/example/client/network/room/RoomService.kt
package com.example.client.network.room

import com.example.client.model.CreateRoom
import com.example.client.model.EditRoom
import com.example.client.model.Room
import com.example.client.utils.ApiResponse
import com.example.client.utils.PagingResponse
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.Path
import retrofit2.http.Query


interface RoomService {
    @POST("api/rooms/")
    suspend fun createRoom(
        @Body createRoom:CreateRoom
    ): ApiResponse<Room>

    @GET("api/rooms/{roomPin}/join")
    suspend fun joinRoom(
        @Path("roomPin") roomPin:String
    ): ApiResponse<Int>

    @GET("api/rooms/{roomPin}/participant")
    suspend fun getRoomForParticipants(
        @Path("roomPin") roomPin:String
    ): ApiResponse<Room>

    @GET("api/rooms/{roomId}")
    suspend fun getRoomDetail(
        @Path("roomId") roomId:Int
    ): ApiResponse<Room>

    @PATCH("api/rooms/{roomId}/end")
    suspend fun endRoom(
        @Path("roomId") roomId:Int
    ): ApiResponse<Boolean>

    @PATCH("api/rooms/")
    suspend fun editRoom(
        @Body editRoom:EditRoom
    ): ApiResponse<Boolean>

    @DELETE("api/rooms/{roomId}")
    suspend fun deleteRoom(
        @Path("roomId") roomId:Int
    ): ApiResponse<Boolean>

    @GET("api/rooms/")
    suspend fun getMyListRoom(
        @Query("keyword") keyword:String,
        @Query("pageIndex") pageIndex:Int,
        @Query("sortBy") sortBy:String = "createdAt",
    ): ApiResponse<PagingResponse<List<Room>>>

    @GET("api/rooms/joined")
    suspend fun getJoinedRoom(
        @Query("pageIndex") pageIndex:Int,
    ): ApiResponse<PagingResponse<List<Room>>>

    @DELETE("api/rooms/{roomId}/kick/{userId}")
    suspend fun kickUser(
        @Path("roomId") roomId:Int,
        @Path("userId") userId:Int
    ): ApiResponse<Boolean>
}